AlphaGo/training/supervised_policy_trainer.py

```diff
@@ -6,276 +6,247 @@
+# Map from transform name to index.
+TRANSFORMATION_INDICES = {name: i for (i, name) in enumerate(sorted(BOARD_TRANSFORMATIONS.keys()))}
 
 
 def one_hot_action(action, size=19):
-    """Convert an (x,y) action into a size x size array of zeros with a 1 at x,y
+    """Helper function to convert an (x, y) action into a size x size array of zeros with a 1 at
+       (x, y).
     """
-
     categorical = np.zeros((size, size))
     categorical[action] = 1
     return categorical
 
 
-class threading_shuffled_hdf5_batch_generator:
-    """A generator of batches of training data for use with the fit_generator function
-       of Keras. Data is accessed in the order of the given indices for shuffling.
+class ShuffledHDF5BatchGenerator:
+    """A generator of batches of training data for use with the fit_generator function of Keras.
+       Data are accessed from the given set of indices.
 
-       it is threading safe but not multiprocessing therefore only use it with
-       pickle_safe=False when using multiple workers
-    """
+       Each row of `indices` must have an index into the dataset in the 0th column and a
+       transform index in the 1st column.
 
-    def shuffle_indices(self, seed=None, idx=0):
-        # set generator_sample to idx
-        self.metadata['generator_sample'] = idx
+       If a `reshuffle_seed` is given, the indices are pseudo-randomly reshuffled aftear each pass
+       through the full dataset. If `initial_batch` is given, then the generator picks up 'where it
+       left off'. If `initial_batch` is greater than the number of batches in the dataset, the
+       indices will be reshuffled `int(initial_batch / batches_per_data_epoch)` times.
 
-        # check if seed is provided or generate random
-        if seed is None:
-            # create random seed
-            self.metadata['generator_seed'] = np.random.random_integers(4294967295)
-
-        # feed numpy.random with seed in order to continue with certain batch
-        np.random.seed(self.metadata['generator_seed'])
-        # shuffle indices according to seed
-        if not self.validation:
-            np.random.shuffle(self.indices)
+       It is threading safe but not multiprocessing. Only use it with `pickle_safe=False` when using
+       multiple workers.
+    """
 
-    def __init__(self, state_dataset, action_dataset, indices, batch_size, metadata=None,
-                 validation=False):
-        self.action_dataset = action_dataset
+    def __init__(self, state_dataset, action_dataset, indices, batch_size, initial_batch=0,
+                 reshuffle_seed=None):
         self.state_dataset = state_dataset
+        self.action_dataset = action_dataset
+        self.n_data = len(state_dataset)
         # lock used for multithreaded workers
-        self.data_lock = threading.Lock()
+        self.lock = threading.Lock()
         self.indices_max = len(indices)
-        self.validation = validation
```

Comment:
`if self.seed is not None` is already included in the functions that call `reshuffle`